File: go/test/tle_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/satellite-tle-data-sdk/go"
	"github.com/voxgig-sdk/satellite-tle-data-sdk/go/core"

	vs "github.com/voxgig-sdk/satellite-tle-data-sdk/go/utility/struct"
)

func TestTleEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Tle(nil)
		if ent == nil {
			t.Fatal("expected non-nil TleEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := tleBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "tle." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set SATELLITETLEDATA_TEST_TLE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		tleRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.tle", setup.data)))
		var tleRef01Data map[string]any
		if len(tleRef01DataRaw) > 0 {
			tleRef01Data = core.ToMapAny(tleRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = tleRef01Data

		// LIST
		tleRef01Ent := client.Tle(nil)
		tleRef01Match := map[string]any{}

		tleRef01ListResult, err := tleRef01Ent.List(tleRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, tleRef01ListOk := tleRef01ListResult.([]any)
		if !tleRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", tleRef01ListResult)
		}

		// LOAD
		tleRef01MatchDt0 := map[string]any{
			"id": tleRef01Data["id"],
		}
		tleRef01DataDt0Loaded, err := tleRef01Ent.Load(tleRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		tleRef01DataDt0LoadResult := core.ToMapAny(tleRef01DataDt0Loaded)
		if tleRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if tleRef01DataDt0LoadResult["id"] != tleRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func tleBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "tle", "TleTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read tle test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse tle test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"tle01", "tle02", "tle03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("SATELLITETLEDATA_TEST_TLE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"SATELLITETLEDATA_TEST_TLE_ENTID": idmap,
		"SATELLITETLEDATA_TEST_LIVE":      "FALSE",
		"SATELLITETLEDATA_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["SATELLITETLEDATA_TEST_TLE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["SATELLITETLEDATA_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewSatelliteTleDataSDK(core.ToMapAny(mergedOpts))
	}

	live := env["SATELLITETLEDATA_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["SATELLITETLEDATA_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/tle_entity_test.rb
# Tle entity test

require "minitest/autorun"
require "json"
require_relative "../SatelliteTleData_sdk"
require_relative "runner"

class TleEntityTest < Minitest::Test
  def test_create_instance
    testsdk = SatelliteTleDataSDK.test(nil, nil)
    ent = testsdk.Tle(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = tle_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "tle." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set SATELLITETLEDATA_TEST_TLE_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    tle_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.tle")))
    tle_ref01_data = nil
    if tle_ref01_data_raw.length > 0
      tle_ref01_data = Helpers.to_map(tle_ref01_data_raw[0][1])
    end

    # LIST
    tle_ref01_ent = client.Tle(nil)
    tle_ref01_match = {}

    tle_ref01_list_result = tle_ref01_ent.list(tle_ref01_match, nil)
    assert tle_ref01_list_result.is_a?(Array)

    # LOAD
    tle_ref01_match_dt0 = {
      "id" => tle_ref01_data["id"],
    }
    tle_ref01_data_dt0_loaded = tle_ref01_ent.load(tle_ref01_match_dt0, nil)
    tle_ref01_data_dt0_load_result = Helpers.to_map(tle_ref01_data_dt0_loaded)
    assert !tle_ref01_data_dt0_load_result.nil?
    assert_equal tle_ref01_data_dt0_load_result["id"], tle_ref01_data["id"]

  end
end

def tle_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "tle", "TleTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = SatelliteTleDataSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["tle01", "tle02", "tle03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["SATELLITETLEDATA_TEST_TLE_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "SATELLITETLEDATA_TEST_TLE_ENTID" => idmap,
    "SATELLITETLEDATA_TEST_LIVE" => "FALSE",
    "SATELLITETLEDATA_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["SATELLITETLEDATA_TEST_TLE_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["SATELLITETLEDATA_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = SatelliteTleDataSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["SATELLITETLEDATA_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["SATELLITETLEDATA_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
